relayer/pkg/chainlink/keys/keystore.go

```diff
@@ -22,4 +58,117 @@ type NonceManager interface {
 	IncrementNextSequence(address caigotypes.Hash, chainID string, currentNonce *big.Int) error
-}
\ No newline at end of file
+}
+
+const (
+	maxPointByteLen = 32 // stark curve max is 252 bits
+	signatureLen    = 2 * maxPointByteLen
+)
+
+// MemKeystore is an in-memory implementation of the LOOPp Keystore interface
+type MemKeystore struct {
+	mu   sync.Mutex
+	keys map[string]*big.Int
+}
+
+func NewMemKeystore() *MemKeystore {
+	return &MemKeystore{
+		keys: make(map[string]*big.Int),
+	}
+}
+
+func (ks *MemKeystore) Put(senderAddress string, k *big.Int) {
+	ks.mu.Lock()
+	defer ks.mu.Unlock()
+	ks.keys[senderAddress] = k
+}
+
+var ErrSenderNoExist = errors.New("sender does not exist")
+
+func (ks *MemKeystore) Get(senderAddress string) (*big.Int, error) {
+	ks.mu.Lock()
+	defer ks.mu.Unlock()
+	k, exists := ks.keys[senderAddress]
+	if !exists {
+		return nil, fmt.Errorf("error getting key for sender %s: %w", senderAddress, ErrSenderNoExist)
+	}
+	return k, nil
+}
+
+// Sign implements the LoopKeystore interface
+// this implementation wraps starknet specific curve and expects
+// hash: byte representation (big-endian) of *big.Int.
+func (ks *MemKeystore) Sign(id string, hash []byte) ([]byte, error) {
+	k, err := ks.Get(id)
+	if err != nil {
+		return nil, err
+	}
+
+	starkHash := new(big.Int).SetBytes(hash)
+
+	x, y, err := caigo.Curve.Sign(starkHash, k)
+	if err != nil {
+		return nil, fmt.Errorf("error signing data with curve: %w", err)
+	}
+
+	s := &signature{
+		x: x,
+		y: y,
+	}
+	return s.bytes()
+}
+
+// signature is an intermediate representation for translating between a raw-bytes signature and a caigo
+// signature comprised of big.Ints
+type signature struct {
+	x, y *big.Int
+}
+
+func (s *signature) bytes() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	n, err := buf.Write(padBytes(s.x.Bytes(), maxPointByteLen))
+	if err != nil {
+		return nil, fmt.Errorf("error writing 'x' component of signature: %w", err)
+	}
+	if n != maxPointByteLen {
+		return nil, fmt.Errorf("unexpected write length of 'x' component of signature: wrote %d expected %d", n, maxPointByteLen)
+	}
+
+	n, err = buf.Write(padBytes(s.y.Bytes(), maxPointByteLen))
+	if err != nil {
+		return nil, fmt.Errorf("error writing 'y' component of signature: %w", err)
+	}
+	if n != maxPointByteLen {
+		return nil, fmt.Errorf("unexpected write length of 'y' component of signature: wrote %d expected %d", n, maxPointByteLen)
+	}
+
+	if buf.Len() != signatureLen {
+		return nil, fmt.Errorf("error in signature length")
+	}
+	return buf.Bytes(), nil
+}
+
+// b is expected to encode x,y components in accordance with [signature.bytes]
+func signatureFromBytes(b []byte) (*signature, error) {
+	if len(b) != signatureLen {
+		return nil, fmt.Errorf("expected slice len %d got %d", signatureLen, len(b))
+	}
+	x := b[:maxPointByteLen]
+	y := b[maxPointByteLen:]
+
+	return &signature{
+		x: new(big.Int).SetBytes(x),
+		y: new(big.Int).SetBytes(y),
```

Comment:
Can these ints be negative? b/c `Bytes()` doesn't preserve the sign. I ran in to this with LOOPs and ended up using a protobuf https://github.com/smartcontractkit/chainlink-relay/blob/main/pkg/loop/internal/pb/big.go

Reply:
that's a good question. 

yes, based on the elliptic curve, they can be negative.

the current OCR key signing for starknet is using big.Int.Bytes, and i largely copied that.

@aalu1418 what is the expectation regarding the sign of the points on the curve -- is the OCR implementation consciously choosing to ignore/drop it, or is that an oversight in the implementation? 

https://github.com/smartcontractkit/chainlink-starknet/blob/develop/relayer/pkg/chainlink/keys/ocr2key.go

Reply:
i looked at the official python implementation. in my reading, all the int values are positive. can one of you read the python and/or caigo give your assessment? cc @archseer 

i'm also going to ask research on the general topic of big.Int.Set/Bytes and elliptic curves.

the private key is positve     https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L103

the msgHash is positive https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L139


the signature components --
r is positive:         https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L156

s is positive: 

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L170

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L111

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/math_utils.py#L52

Reply:
yes, good idea. i think the right place for that is the `Decode` implementation, which i've now added

Reply:
Where is the sanity check? I see `(s *signature) bytes()` checking length, but not sign.

Reply:
If `(s *signature) bytes()` is called with a signature that has negative values, it will silently drop that information. IMHO this is where the panic should occur: at the narrowest scope without making assumptions about anything else.